sw/device/tests/otbn/otbn_rsa_test.c

```diff
@@ -0,0 +1,903 @@
+    return kOtbnRsaError;
+  }
+
+  // Call OTBN to perform operation
+  otbn_rsa_result_t rv;
+  rv = otbn_call_function(ctx, &kOtbnFuncRsaRsaEncrypt);
+  if (rv != kOtbnRsaOk) {
+    return rv;
+  }
+
+  rv = otbn_wait_for_done(ctx);
+  if (rv != kOtbnRsaOk) {
+    return rv;
+  }
+
+  if (dif_otbn_dmem_read(&ctx->dif, ctx->otbnptr_out, out, size_bytes) !=
+      kDifOtbnOk) {
+    return kOtbnRsaError;
+  }
+
+  return kOtbnRsaOk;
+}
+
+/**
+ * Decrypt a message with RSA
+ *
+ * @param ctx the context object
+ * @param modulus the modulus n
+ * @param private_exponent the private exponent (d)
+ * @param in the encrypted message
+ * @param out the decrypted (plaintext) message
+ * @param size_bytes the size of all buffers in bytes, i.e. the key/modulus
+ *                   length (i.e. 128 for RSA 1024). Valid range: 32..512 in
+ *                   32 byte-steps (i.e. RSA 256 to RSA 4096).
+ * @return kOtbnRsaOk if the operation was successful, kOtbnRsaError otherwise
+ */
+static otbn_rsa_result_t otbn_rsa_decrypt(otbn_rsa_t *ctx,
+                                          const uint8_t *modulus,
+                                          const uint8_t *private_exponent,
+                                          const uint8_t *in, uint8_t *out,
+                                          size_t size_bytes) {
+  if (ctx == NULL || !ctx->is_loaded) {
+    return kOtbnRsaBadArg;
+  }
+  if (size_bytes % 32 != 0) {
+    return kOtbnRsaBadArg;
+  }
+  uint32_t n_limbs = size_bytes / 32;
+  if (n_limbs == 0 || n_limbs > 16) {
+    return kOtbnRsaBadArg;
+  }
+
+  // Write input arguments
+  if (dif_otbn_dmem_write(&ctx->dif, ctx->otbnptr_n_limbs, &n_limbs, 4) !=
+      kDifOtbnOk) {
+    return kOtbnRsaError;
+  }
+
+  if (dif_otbn_dmem_write(&ctx->dif, ctx->otbnptr_modulus, modulus,
+                          size_bytes) != kDifOtbnOk) {
+    return kOtbnRsaError;
+  }
+
+  if (dif_otbn_dmem_write(&ctx->dif, ctx->otbnptr_exp, private_exponent,
+                          size_bytes) != kDifOtbnOk) {
+    return kOtbnRsaError;
+  }
+
+  if (dif_otbn_dmem_write(&ctx->dif, ctx->otbnptr_in, in, size_bytes) !=
+      kDifOtbnOk) {
+    return kOtbnRsaError;
+  }
+
+  // Call OTBN to perform operation
+  otbn_rsa_result_t rv;
+  rv = otbn_call_function(ctx, &kOtbnFuncRsaRsaDecrypt);
+  if (rv != kOtbnRsaOk) {
+    return rv;
+  }
+
+  rv = otbn_wait_for_done(ctx);
+  if (rv != kOtbnRsaOk) {
+    return rv;
+  }
+
+  // Read back result
+  if (dif_otbn_dmem_read(&ctx->dif, ctx->otbnptr_out, out, size_bytes) !=
+      kDifOtbnOk) {
+    return kOtbnRsaError;
+  }
+
+  return kOtbnRsaOk;
+}
+
+static void check_data(const uint8_t *actual, const uint8_t *expected,
+                       size_t size_bytes) {
+  for (int i = 0; i < size_bytes; ++i) {
+    CHECK(actual[i] == expected[i],
+          "Data mismatch at byte %d: 0x%x (actual) != 0x%x (expected)", i,
+          actual[i], expected[i]);
+  }
+}
+
+uint64_t t_start;
+uint64_t t_end;
```

Comment:
These should be static; also, why are they globals?

Reply:
They are global for convenience. It's just a profiling helper in a test.